src/textual/widgets/tabs.py

```diff
@@ -0,0 +1,350 @@
+from textual.keys import Keys
+from textual.reactive import Reactive
+from textual.renderables.opacity import Opacity
+from textual.renderables.underline_bar import UnderlineBar
+from textual.widget import Widget
+
+__all__ = ["Tab", "Tabs"]
+
+
+@dataclass
+class Tab:
+    """Data container representing a single tab.
+
+    Attributes:
+        label (str): The user-facing label that will appear inside the tab.
+        name (str, optional): A unique string key that will identify the tab. If None, it will default to the label.
+            If the name is not unique within a single list of tabs, only the final Tab will be displayed.
+    """
+
+    label: str
+    name: str | None = None
+
+    def __post_init__(self):
+        if self.name is None:
+            self.name = self.label
+
+    def __str__(self):
+        return self.label
+
+
+class TabsRenderable:
+    """Renderable for the Tabs widget."""
+
+    def __init__(
+        self,
+        tabs: Iterable[Tab],
+        *,
+        active_tab_name: str,
+        active_tab_style: StyleType,
+        active_bar_style: StyleType,
+        inactive_tab_style: StyleType,
+        inactive_bar_style: StyleType,
+        inactive_text_opacity: float,
+        tab_padding: int | None,
+        bar_offset: float,
+        width: int | None = None,
+    ):
+        self.tabs = {tab.name: tab for tab in tabs}
+
+        try:
+            self.active_tab_name = active_tab_name or next(iter(self.tabs))
+        except StopIteration:
+            self.active_tab_name = None
+
+        self.active_tab_style = active_tab_style
+        self.active_bar_style = active_bar_style
+
+        self.inactive_tab_style = inactive_tab_style
+        self.inactive_bar_style = inactive_bar_style
+
+        self.bar_offset = bar_offset
+        self.width = width
+        self.tab_padding = tab_padding
+        self.inactive_text_opacity = inactive_text_opacity
+
+        self._label_range_cache: dict[str, tuple[int, int]] = {}
+        self._selection_range_cache: dict[str, tuple[int, int]] = {}
+
+    def __rich_console__(
+        self, console: Console, options: ConsoleOptions
+    ) -> RenderResult:
+        if self.tabs:
+            yield from self.get_tab_labels(console, options)
+        yield Segment.line()
+        yield from self.get_underline_bar(console)
+
+    def get_tab_labels(self, console: Console, options: ConsoleOptions) -> RenderResult:
+        """Yields the spaced-out labels that appear above the line for the Tabs widget"""
+        width = self.width or options.max_width
+        tab_values = self.tabs.values()
+
+        space = cast(
+            Edge, SimpleNamespace(min_size=1, size=self.tab_padding or None, fraction=1)
+        )
+        edges = []
+        for tab in tab_values:
+            tab = cast(
+                Edge,
+                SimpleNamespace(size=cell_len(tab.label), min_size=1, fraction=None),
+            )
+            edges.extend([space, tab, space])
```

Comment:
I get why there is `[space, tab, space]` to generate space around tabs.

This would give you something like: `[space, tab, space, space, tab, space]` etc. for multiple tabs.

It would be more efficient for layout resolve if you combined those double spaces in to a single space with `fraction=2`, which should be equivalent. So you would generate the following:

`[space, tab, double_space, tab, space]`

Reply:
As discussed, have kept it as is in order to have layout_resolve tell us the extents of each tab without further work.

Reply:
I see why you might need the repetition to properly define the clickable boundaries. I withdraw my objection!